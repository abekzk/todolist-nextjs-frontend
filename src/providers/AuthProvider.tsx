import User from '../models/user';
import {
  getUser,
  loginUser,
  logoutUser,
  createUser,
} from '../services/firebase/firebase';
import {
  ReactNode,
  useContext,
  useEffect,
  createContext,
  useState,
} from 'react';

interface AuthState {
  user?: User | null;
}

interface AuthContextType {
  auth: AuthState;
  signUp: (email: string, password: string) => Promise<void>;
  signIn: (email: string, password: string) => Promise<void>;
  signOut: () => Promise<void>;
}

const AuthContext = createContext({} as AuthContextType);

export const AuthProvider = ({ children }: { children?: ReactNode }) => {
  const [auth, setAuth] = useState<AuthState>({});

  async function signUp(email: string, password: string) {
    try {
      const user = await createUser(email, password);
      setAuth({ user: user });
    } catch (err) {
      throw err;
    }
  }

  async function signIn(email: string, password: string) {
    try {
      const user = await loginUser(email, password);
      setAuth({ user: user });
    } catch (err) {
      throw err;
    }
  }

  async function signOut() {
    try {
      await logoutUser();
      setAuth({ user: null });
    } catch (err) {
      throw err;
    }
  }

  useEffect(() => {
    (async () => {
      try {
        const user = await getUser();
        setAuth({ user: user });
      } catch (err) {
        setAuth({ user: null });
      }
    })();
  }, []);

  return (
    <AuthContext.Provider value={{ auth, signUp, signIn, signOut }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => useContext(AuthContext);
