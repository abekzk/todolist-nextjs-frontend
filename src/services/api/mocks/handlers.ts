import { API_URL } from '../../../configs/config';
import task from './data/task.json';
import tasks from './data/tasks.json';
import { rest } from 'msw';

export const handlers = [
  rest.get(API_URL + '/v1/tasks', (_, res, ctx) => {
    return res(ctx.json(tasks));
  }),

  rest.post(API_URL + '/v1/tasks', (_, res, ctx) => {
    return res(ctx.status(201), ctx.json(task));
  }),

  rest.put(API_URL + '/v1/tasks/:taskId', (_, res, ctx) => {
    return res(ctx.status(200), ctx.json(task));
  }),

  rest.delete(API_URL + '/v1/tasks/:taskId', (_, res, ctx) => {
    return res(ctx.status(204));
  }),
];
