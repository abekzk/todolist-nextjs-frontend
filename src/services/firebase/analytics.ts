import { getAnalytics, logEvent as log } from 'firebase/analytics';

// タスク関連イベントシグネチャ
function logEvent(key: 'add_task'): void;
function logEvent(key: 'update_task'): void;
function logEvent(key: 'delete_task'): void;

// ログイン関連イベントシグネチャ
function logEvent(key: 'login', value: { method: 'email' | 'google' }): void;
function logEvent(key: 'sign_up', value: { method: 'email' | 'google' }): void;
function logEvent(key: 'logout'): void;

// イベントログの処理
function logEvent(key: string, value?: object) {
  if (typeof window == 'undefined') return; // レンダリング後のみ有効

  const analytics = getAnalytics();
  log(analytics, key, value);
}

// ビューログの処理
function logView(contentType: string, params?: object) {
  if (typeof window == 'undefined') return; // レンダリング後のみ有効

  const analytics = getAnalytics();
  log(analytics, 'view_content', {
    content_type: contentType,
    ...params,
  });
}

// page_view用のパラメータ
type PageParams = {
  page_title?: string;
  page_location?: string;
  page_path?: string;
};

// ページビューログの処理
function logPage(params: PageParams) {
  if (typeof window == 'undefined') return; // レンダリング後のみ有効

  const analytics = getAnalytics();
  log(analytics, 'page_view', params);
}

export { logEvent, logView, logPage };
