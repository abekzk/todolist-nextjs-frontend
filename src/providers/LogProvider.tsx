import { logPage } from '../services/firebase/analytics';
import { useRouter } from 'next/router';
import { ReactNode, useEffect, createContext } from 'react';

const LogContext = createContext(null);

export const LogProvider = ({ children }: { children?: ReactNode }) => {
  const router = useRouter();

  useEffect(() => {
    const handleRouteChange = () => {
      logPage({
        page_location: location.href,
        page_title: document?.title,
      });
    };
    router.events.on('routeChangeComplete', handleRouteChange);
    return () => {
      router.events.off('routeChangeComplete', handleRouteChange);
    };
  }, [router.events]);

  return <LogContext.Provider value={null}>{children}</LogContext.Provider>;
};
