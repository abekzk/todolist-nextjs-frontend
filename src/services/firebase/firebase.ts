import { FIREBASE_CONFIG } from '../../configs/config';
import User from '../../models/user';
import { initializeAnalytics } from 'firebase/analytics';
import { initializeApp, getApps } from 'firebase/app';
import Firebase, {
  getAuth,
  onAuthStateChanged,
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signOut,
} from 'firebase/auth';

export function initFirebase() {
  if (typeof window != 'undefined' && !getApps().length) {
    // Initialize Firebase
    const app = initializeApp(FIREBASE_CONFIG);
    // Initialize Analytics and get a reference to the service
    initializeAnalytics(app);
  }
}

function toModelUser(_user: Firebase.UserInfo): User {
  const user: User = {
    email: _user.email ?? '',
  };
  return user;
}

// Auth情報取得
async function getFirebaseAuthUser(): Promise<Firebase.User> {
  const auth = getAuth();
  return new Promise((resolve, reject) => {
    onAuthStateChanged(auth, (user) => {
      if (user) {
        resolve(user);
      }
      reject(new Error('No Login'));
    });
  });
}

// ユーザー取得
export async function getUser(): Promise<User> {
  try {
    const user = await getFirebaseAuthUser();
    return toModelUser(user);
  } catch (err) {
    throw new Error('getUser Error');
  }
}

// トークン取得
export async function getToken(): Promise<string> {
  try {
    const user = await getFirebaseAuthUser();
    const token = await user.getIdToken();
    return token;
  } catch (err) {
    throw new Error('getToken Error');
  }
}

// ログイン
export async function loginUser(
  email: string,
  password: string
): Promise<User> {
  const auth = getAuth();
  try {
    const userCredential = await signInWithEmailAndPassword(
      auth,
      email,
      password
    );
    return toModelUser(userCredential.user);
  } catch (err) {
    throw new Error('loginUser Error');
  }
}

// ログアウト
export async function logoutUser() {
  const auth = getAuth();
  try {
    await signOut(auth);
  } catch (err) {
    throw new Error('logoutUser Error');
  }
}

// ユーザー登録
export async function createUser(
  email: string,
  password: string
): Promise<User> {
  const auth = getAuth();
  try {
    const userCredential = await createUserWithEmailAndPassword(
      auth,
      email,
      password
    );
    return toModelUser(userCredential.user);
  } catch (err) {
    throw new Error('createUser Error');
  }
}
