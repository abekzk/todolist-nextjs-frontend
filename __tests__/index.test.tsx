import IndexPage from '../src/pages/index';
import { AuthProvider } from '../src/providers/AuthProvider';
import { mockServer } from '../src/services/api/mocks/server';
import { getToken, getUser } from '../src/services/firebase/authentication';
import { render, screen, waitFor } from '@testing-library/react';

// モック
jest.mock('../src/services/firebase/authentication');
const mockGetToken = getToken as jest.MockedFunction<typeof getToken>;
mockGetToken.mockResolvedValue('testtoken');
const mockGetUser = getUser as jest.MockedFunction<typeof getUser>;
mockGetUser.mockResolvedValue({ email: 'test@example.com' });
jest.mock('../src/services/firebase/analytics');

beforeAll(() => mockServer.listen());
afterEach(() => mockServer.resetHandlers());
afterAll(() => mockServer.close());

test('renders IndexPage', async () => {
  const { container } = render(
    <AuthProvider>
      <IndexPage />
    </AuthProvider>
  );
  await waitFor(() => screen.getByText('タスク名1'));
  expect(container).toMatchSnapshot();
});
