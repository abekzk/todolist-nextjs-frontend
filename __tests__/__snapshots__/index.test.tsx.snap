// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders IndexPage 1`] = `
<div>
  <div>
    <header>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorDefault MuiAppBar-positionStatic css-1gz3byy-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t018ql-MuiToolbar-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-uh02xs-MuiTypography-root"
          >
            TODOリストアプリ
          </h6>
          <button
            aria-controls="menu-appbar"
            aria-haspopup="true"
            aria-label="account of current user"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AccountCircleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </header>
    </header>
    <main>
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthSm css-1snf7pn-MuiContainer-root"
        >
          <form
            novalidate=""
          >
            <div
              class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for=":r0:"
                id=":r0:-label"
              >
                タスクを入力
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  name="title"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-1in441m"
                  >
                    <span>
                      タスクを入力
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AddCircleOutlineRoundedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 7c-.55 0-1 .45-1 1v3H8c-.55 0-1 .45-1 1s.45 1 1 1h3v3c0 .55.45 1 1 1s1-.45 1-1v-3h3c.55 0 1-.45 1-1s-.45-1-1-1h-3V8c0-.55-.45-1-1-1zm0-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
                  />
                </svg>
              </span>
              Add Todo
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </form>
        </div>
        <div
          class="MuiContainer-root MuiContainer-maxWidthMd css-11ktbtp-MuiContainer-root"
        >
          <ul
            class="MuiList-root MuiList-padding MuiList-dense css-h4y409-MuiList-root"
          >
            <li
              class="MuiListItem-container css-oyakqy-MuiListItem-container"
            >
              <div
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding MuiListItem-secondaryAction css-1ob1ktf-MuiListItem-root"
              >
                <div
                  class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                >
                  <span
                    class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root PrivateSwitchBase-edgeStart css-8l0d08-MuiButtonBase-root-MuiCheckbox-root"
                  >
                    <input
                      class="PrivateSwitchBase-input css-1m9pwf3"
                      data-indeterminate="false"
                      tabindex="-1"
                      type="checkbox"
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="CheckBoxOutlineBlankIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </span>
                </div>
                <div
                  class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-70yyfe-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
                  >
                    タスク名1
                  </span>
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                  >
                    説明文1
                  </p>
                </div>
              </div>
              <div
                class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
              >
                <button
                  aria-label="Edit"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="delete"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="DeleteOutlineRoundedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V9c0-1.1-.9-2-2-2H8c-1.1 0-2 .9-2 2v10zM9 9h6c.55 0 1 .45 1 1v8c0 .55-.45 1-1 1H9c-.55 0-1-.45-1-1v-8c0-.55.45-1 1-1zm6.5-5-.71-.71c-.18-.18-.44-.29-.7-.29H9.91c-.26 0-.52.11-.7.29L8.5 4H6c-.55 0-1 .45-1 1s.45 1 1 1h12c.55 0 1-.45 1-1s-.45-1-1-1h-2.5z"
                    />
                  </svg>
                </button>
              </div>
            </li>
            <li
              class="MuiListItem-container css-oyakqy-MuiListItem-container"
            >
              <div
                class="MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiListItem-padding MuiListItem-secondaryAction css-1ob1ktf-MuiListItem-root"
              >
                <div
                  class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
                >
                  <span
                    class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked PrivateSwitchBase-edgeStart css-8l0d08-MuiButtonBase-root-MuiCheckbox-root"
                  >
                    <input
                      checked=""
                      class="PrivateSwitchBase-input css-1m9pwf3"
                      data-indeterminate="false"
                      tabindex="-1"
                      type="checkbox"
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="CheckBoxIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                      />
                    </svg>
                  </span>
                </div>
                <div
                  class="MuiListItemText-root MuiListItemText-dense MuiListItemText-multiline css-1x6r76p-MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
                  >
                    タスク名2
                  </span>
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
                  >
                    説明文2
                  </p>
                </div>
              </div>
              <div
                class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
              >
                <button
                  aria-label="Edit"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="EditIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="delete"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="DeleteOutlineRoundedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V9c0-1.1-.9-2-2-2H8c-1.1 0-2 .9-2 2v10zM9 9h6c.55 0 1 .45 1 1v8c0 .55-.45 1-1 1H9c-.55 0-1-.45-1-1v-8c0-.55.45-1 1-1zm6.5-5-.71-.71c-.18-.18-.44-.29-.7-.29H9.91c-.26 0-.52.11-.7.29L8.5 4H6c-.55 0-1 .45-1 1s.45 1 1 1h12c.55 0 1-.45 1-1s-.45-1-1-1h-2.5z"
                    />
                  </svg>
                </button>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </main>
    <footer />
  </div>
</div>
`;
